### PR TITLE
Exporting more Hash related functions

### PR DESCRIPTION
**Pull request by cmarcelo**
Following new functionality is exported:
- Calculate swupd hashes without having a file on the disk
- Calculate swupd hashes without having a the whole data in memory

Exports a type containing exactly the metadata used by the swupd hash.

Doesn't change anything in the Hashval or interning, everything should work as before.

### CODE REVIEW

**Comment by cmarcelo** [comment]
There isn't an use case yet in swupd content generation. This is more useful for "client" type applications. It could also help smooth things for changes to support reading the UID/GID information from other sources.

I have non swupd-server code that uses it to validate the tarball before writing the extracted contents to disk. Would be useful to have it part of the swupd package, but I'm fine closing this and waiting until we have an official usage of this functionality.

File: swupd/hash.go [hunk 1]
```diff
@@ -241,3 +173,67 @@ func (h *Hash) Sum() string {
...
+	if err != nil {
+		return "", fmt.Errorf("error creating hash for file %s: %s", filename, err)
+	}
+
+	if info.Mode&syscall.S_IFMT == syscall.S_IFREG {
+		f, err := os.Open(filename)
+		if err != nil {
+			return "", fmt.Errorf("read error for file %s: %s", filename, err)
+		}
+		defer func() {
```
Comment (icarus-sparry) [inline swupd/hash.go]: My usual comment about not deferring something for 1 line.

File: swupd/hash.go [hunk 2]
```diff
@@ -241,3 +173,67 @@ func (h *Hash) Sum() string {
...
+
+	hashInfo := &HashFileInfo{
+		Mode: info.Mode,
+		UID:  info.Uid,
+		GID:  info.Gid,
+		Size: info.Size,
+	}
+
+	if info.Mode&syscall.S_IFMT == syscall.S_IFLNK {
+		fmt.Printf("\n\n\n%s  %x  %x  %x\n\n\n\n", filename, info.Mode, syscall.S_IFLNK, info.Mode&syscall.S_IFLNK)
```
Comment (icarus-sparry) [inline swupd/hash.go]: Debugging Printf ???

Comment (cmarcelo) [reply swupd/hash.go]: Haha, sorry for that :-)  Fixed. Was learning how to use S_IFMT.

File: swupd/hash.go (hunk 1 above)
Comment (cmarcelo) [reply swupd/hash.go]: Done.

**Comment by cmarcelo** [comment]
Pushed again to trigger travis.

File: swupd/hash.go [hunk 3]
```diff
@@ -241,3 +173,63 @@ func (h *Hash) Sum() string {
 	hex.Encode(result[:], h.hmac.Sum(nil))
 	return string(result[:])
 }
+
+// GetHashFile calculate the swupd hash for a file in the disk.
+func GetHashFile(filename string) (string, error) {
```
Comment (tmarcu) [inline swupd/hash.go]: Sorry if I'm being too picky...but GetHashFile sounds like it's returning me a - Hash "file", not a string. Would it make more sense to just call it GetHash? or GetFileHash()?

Comment (cmarcelo) [reply swupd/hash.go]: Maybe GetHashForFile and GetHashForBytes?

Comment (icarus-sparry) [reply swupd/hash.go]: @tmarcu You can (almost) never be too picky about exported Names

File: swupd/hash.go [hunk 4]
```diff
@@ -134,8 +135,109 @@ func hmacComputeKey(info syscall.Stat_t) []byte {
...
+	}
+	hex.Encode(key[:], mac.Sum(nil))
+
+	// With the key in hand, create the HMAC struct that will be
+	// used to write data to.
+	h := &Hash{
+		hmac: hmac.New(sha256.New, key[:]),
+	}
+
+	// Pre-write data we known so that directories and symbolic
```
Comment (tmarcu) [inline swupd/hash.go]: s/known/know

**Pull request merged**